### ISSUE TITLE
Coab is failing to retry failed push commits upon concurrency failure

### ISSUE THREAD

**Issue by gberche-orange** [orange-cloudfoundry/paas-templates#1223]
### Expected behavior

* As a marketplace user 
* In order to provision coab service instances
* I need coab to reliably accept OSB calls and save them to git even in case of race conditions on the secret repo

### Observed behavior
 
Coab seems to fail to git push on the secret repo and does not try to pull rebase.

Note that in the case of a failure to server an OSB provisionning request, the cloudfoundry client will issue a delete to avoid orphan.

However, in the case of a failure to server an OSB unprovisionning request, this might head to an inconsistency between paas-secret/paas-templates repo and lock the coa pipelines

```
92b37c8b87ae03a40a6cf126ea5 1623947227 -----p
2021-06-17T16:27:07.63+0000 [APP/PROC/WEB/0] OUT 2021-06-17 16:27:07.638  INFO 8 --- [io-8080-exec-10] c.o.o.c.b.o.o.git.SimpleGitManager       : [paas-secrets.] Preparing to push: behind commits=0 ahead_commits=1
2021-06-17T16:27:07.63+0000 [APP/PROC/WEB/0] OUT 2021-06-17 16:27:07.638  INFO 8 --- [io-8080-exec-10] c.o.o.c.b.o.o.git.SimpleGitManager       : [paas-secrets.] pushing to master ...
2021-06-17T16:27:08.32+0000 [APP/PROC/WEB/0] OUT 2021-06-17 16:27:08.327  INFO 8 --- [io-8080-exec-10] c.o.o.c.b.o.o.git.SimpleGitManager       : [paas-secrets.] pushed ...
2021-06-17T16:27:08.32+0000 [APP/PROC/WEB/0] OUT 2021-06-17 16:27:08.328 ERROR 8 --- [io-8080-exec-10] c.o.o.c.b.o.o.git.SimpleGitManager       : [paas-secrets.] unrecoverable push failure: [REJECTED_O
THER_REASON] with full details: org.eclipse.jgit.transport.PushResult@661012[remoteUpdates={refs/heads/master=RemoteRefUpdate[remoteName=refs/heads/master, REJECTED_OTHER_REASON, 6ebf3085bd7765aa4e66bfd5b
83c73659cbf939b...4929ac06033d292b37c8b87ae03a40a6cf126ea5, fastForward, srcRef=refs/heads/master, message="failed to update ref"]},advertisedRefs={refs/heads/master=Ref[refs/heads/master=6ebf3085bd7765aa
4e66bfd5b83c73659cbf939b], refs/tags/42.0.4=Ref[refs/tags/42.0.4=73e5add38d1deeec37bd3ddbc0f9ac7160dbfb70], refs/tags/44.0.5=Ref[refs/tags/44.0.5=7ee44ea61e67b5b386adf26059774fa33f712385], refs/tags/46.0.
0=Ref[refs/tags/46.0.0=9a657ecd1750650b182aa052d8dcd636cd07d812], refs/tags/46.7.1=Ref[refs/tags/46.7.1=a7c695deed330370319b13aa7cc02f9e4407e2c1], refs/tags/47.0.1=Ref[refs/tags/47.0.1=e01ae72fc9b656e0dc3
708340204170c28766b3e], refs/tags/47.0.2=Ref[refs/tags/47.0.2=254df753234b8831512b370d17ca30035d62d22a], refs/tags/47.0.3=Ref[refs/tags/47.0.3=cad6ee6340928e26693e6b1b2cfcd28fd3625736], refs/tags/48.0.2=R
ef[refs/tags/48.0.2=19b43da31db7726540794b9f3ad247c395279322], refs/tags/51.0.0=Ref[refs/tags/51.0.0=fb2a4446860fc2831f3a13bcbbd9056a2f2dbdfe], refs/tags/51.0.1=Ref[refs/tags/51.0.1=ceec3a5230dfde1051b26e
90088f2f02ef355d95], refs/tags/51.0.2=Ref[refs/tags/51.0.2=90f9a5123a3bd86bbc04315e3b1642ec916ad4be], refs/tags/51.0.3=Ref[refs/tags/51.0.3=926c0d2237b2ee2029320b3f058f088f277967a7], refs/tags/end-bootstr
ap=Ref[refs/tags/end-bootstrap=0130a253292d331906c4a0115b48c5e052a6861e]},uri=https://gitlab-gitlab-k8s.redacted-domain.org/paas_templates_group/paas-templates-secrets.git,updates={refs/
remotes/origin/master=TrackingRefUpdate[refs/heads/master -> refs/remotes/origin/master (forced) 6ebf308..4929ac0]},messageBuffer=error: cannot lock ref 'refs/heads/master': is at a5a5fd1d7547d5c2d09a45a7
da88d1175f498388 but expected 6ebf3085bd7765aa4e66bfd5b83c73659cbf939b
2021-06-17T16:27:08.32+0000 [APP/PROC/WEB/0] OUT ,peerUserAgent=git/2.29.0]
2021-06-17T16:27:08.33+0000 [APP/PROC/WEB/0] OUT 2021-06-17 16:27:08.331  WARN 8 --- [io-8080-exec-10] c.o.o.c.b.o.o.git.SimpleGitManager       : [paas-secrets.] caught java.lang.RuntimeException: unre
coverable push failure : failedStatuses=[REJECTED_OTHER_REASON]
2021-06-17T16:27:08.33+0000 [APP/PROC/WEB/0] OUT java.lang.RuntimeException: unrecoverable push failure : failedStatuses=[REJECTED_OTHER_REASON]
2021-06-17T16:27:08.33+0000 [APP/PROC/WEB/0] OUT     at com.orange.oss.cloudfoundry.broker.opsautomation.ondemandbroker.git.SimpleGitManager.pushCommits(SimpleGitManager.java:433) ~[cf-ops-automation-b
roker-core-1.5.0-SNAPSHOT.jar:1.5.0-SNAPSHOT]
```

With associated commit hashes:

```
$ git show a5a5fd1d7547d5c2d09a45a7da88d1175f498388
commit a5a5fd1d7547d5c2d09a45a7da88d1175f498388
Author: coa-cf-rabbit-broker <redacted-email@redacted-domain.org>
Date:   Thu Jun 17 16:27:06 2021 +0000

    cf-rabbit broker: delete instance id=9aa75d05-b8b4-4127-ad49-7f999c9851c6

    Requested by user_guid=8eb11998-acd3-47cf-a7ac-473ac300cdb4

diff --git a/coab-depls/r_9aa75d05-b8b4-4127-ad49-7f999c9851c6/enable-deployment.yml b/coab-depls/r_9aa75d05-b8b4-4127-ad49-7f999c9851c6/enable-deployment.yml
deleted file mode 100644
index ed97d539..00000000
--- a/coab-depls/r_9aa75d05-b8b4-4127-ad49-7f999c9851c6/enable-deployment.yml
+++ /dev/null


$ git show 6ebf3085bd7765aa4e66bfd5b83c73659cbf939b
commit 6ebf3085bd7765aa4e66bfd5b83c73659cbf939b
Author: coa-cf-mysql-broker <redacted-email@redacted-domain.org>
Date:   Thu Jun 17 16:26:58 2021 +0000

    cf-mysql broker: create instance id=b371b86f-379b-471c-b2f7-76d8861aec96
    
    Requested from space_guid=66d692ed-420b-46da-8054-6aa82dd0c3af org_guid=93c5f5ad-1718-46b7-9af3-d00499945cb2 by user_guid=8eb11998-acd3-47cf-a7ac-473ac300cdb4

diff --git a/coab-depls/y_b371b86f-379b-471c-b2f7-76d8861aec96/enable-deployment.yml b/coab-depls/y_b371b86f-379b-471c-b2f7-76d8861aec96/enable-deployment.yml
new file mode 100644
index 00000000..ed97d539
--- /dev/null
+++ b/coab-depls/y_b371b86f-379b-471c-b2f7-76d8861aec96/enable-deployment.yml
```

### Affected releases

* paas-templates 51.0.3
  * gitlab  13.6.1 k8s
* coab  patched

**Comment by gberche-orange** [comment]
The observed symptom seems observed in a git push race condition according to the following sources, and is likely unrelated to switch from gitlab bosh to gitlab k8s, but more related to concurrency condition triggered by usage of the `main/micro-depls-retrigger-all-deployments/manual-enable-coab-read-only-mode` followed with `main/micro-depls-retrigger-all-deployments/manual-disable-coab-read-only-mode`. Consequently, coab should attempt to rebase and push upon this symptom.

http://git.661346.n2.nabble.com/Git-push-race-condition-td7606412.html

The same symptom may also occasionally be observed with a clone having a stable copy of the remote ref (possibly following an history rewrite, but which is not the case in paas-templates for the secret repo.)

Some more related background
* https://stackoverflow.com/a/19964392/1484823 git locking mechanism to ensure concurrency
* https://linuxpip.org/git-error-cannot-lock-ref/ describes this symptom in various scenarios and solution when the problem is persistent across push retries
* https://git-scm.com/docs/git-update-ref describes the steps for updating a ref including atomic locking and rollback

This error message is part of the automatical test suite within the git repo at https://github.com/git/git/blob/670b81a890388c60b7032a4f5b879f2ece8c4558/t/t1404-update-ref-errors.sh#L342-L350 with a test labelled as `incorrect old value blocks update`

**Comment by gberche-orange** [comment]
fixed in v52 with [coab 1.5.0](https://github.com/orange-cloudfoundry/cf-ops-automation-broker/releases/tag/v1.5.0) by https://github.com/orange-cloudfoundry/cf-ops-automation-broker/commit/3d348c79eec7a97651f9eebc9779b92760109aa1
